### PR TITLE
[RFR] Watch a game from another player

### PR DESCRIPTION
**Pull request by Luwangel**
- [x] Set a token to identify a game's owner
- [x] Setup an authentication based on cookies
- [x] Display a non movable grid to the visitors
- [x] See the moves in real time

### CODE REVIEW

File: app/src/Controller/GameController.php [hunk 1]
```diff
@@ -5,53 +5,77 @@
...
+  }
+
+  private function removeAccess($response) {
+    $response->headers->clearCookie(self::COOKIE_NAME);  
+  }
+
+  private function renderGrid(string $id, Game $game, $hasAccess) {
     return new Response($this->twig->render('game.html.twig', [
       'id' => $id,
       'grid' => $game->getCurrentGrid(),
```
Comment (jdemangeon) [inline app/src/Controller/GameController.php]: You can directly pass the game (or GameContext) object to the template.

File: app/src/Controller/GameController.php [hunk 2]
```diff
@@ -5,53 +5,77 @@
...
 
   public function __construct(\Twig_Environment $twig, GameApi $gameApi, GameRepository $gameRepository) {
-    $this->api = $gameApi;
     $this->twig = $twig;
+    $this->api = $gameApi;
     $this->gameRepository = $gameRepository;
   }
 
-  private function renderGrid(string $id, Game $game) {
+  private function hasAccess($request, $game) {
```
Comment (jdemangeon) [inline app/src/Controller/GameController.php]: It's work but it's a little tricky.. Moreover, private functions are difficult to test.

As I said at the office, you should use the Symfony auth / role system (which is heavy for this project)... Or parameter resolving as described bellow.

https://symfony.com/doc/master/controller/argument_value_resolver.html

This system let you create a class / service which is responsible of the parameter resolution of your controller method.

For example, you can trigger it when a "GameContext" object is required in the controller method signature and retrieve the game with the access status from the ID in the URL and the cookie.

Let's discuss it tomorrow ;)

File: app/src/Entity/Game.php [hunk 3]
```diff
@@ -35,8 +42,9 @@ class Game {
    * @ORM\Column(type="boolean")
    */
   private $isVictory;
-  
+
   public function __construct($resolvedGrid, $currentGrid, $isVictory=false) {
+    $this->token = bin2hex(random_bytes(self::TOKEN_LENGTH));
```
Comment (jdemangeon) [inline app/src/Entity/Game.php]: Domain objects needs to be "neutral". They must not contains any "logic". I'll create this token in the "new" method of the API where the object is instancied for the first time and give him the value with "setToken".

File: app/templates/game.html.twig [hunk 4]
```diff
@@ -1,13 +1,23 @@
 {% extends "base.html.twig" %}
 
-{% block header %}   
+{% block refresh %}
+  {% if not isOwner and not game.isVictory %}
+    <meta http-equiv="refresh" content="1">
```
Comment (alexisjanvier) [inline app/templates/game.html.twig]: :) i didn't know that .... 
but isn't every second a little bit much?

File: app/src/Api/Token.php [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace App\Api;
+
+class Token {
+  public const TOKEN_LENGTH = 10;
+
+  public static function generate(int $tokenLength = self::TOKEN_LENGTH) : string {
```
Comment (alexisjanvier) [inline app/src/Api/Token.php]: Can't you declare `trait Token` rather than a using static method?

**Review by alexisjanvier — APPROVED** [review]

**Pull request merged**